Add new example Ultralytics `YOLO-Interactive-Tracking-UI`

alireza787b opened the pull request:
## ✨ Summary

This PR adds a new educational and interactive example under `examples/YOLO-Interactive-Tracking-UI/`.

The example provides a real-time object tracking interface using Ultralytics YOLO (via NCNN format) and OpenCV. It allows users to click on detected objects to initiate tracking, visualize active tracks with scope overlays and bounding boxes, and dynamically inspect tracking information through terminal logs.

---

## 📂 What's Included

```text
YOLO-Interactive-Tracking-UI/
├── yolo/                    # Stores YOLO models (.pt and exported NCNN)
├── add_yolo_model.py        # Download + export models to NCNN
├── interactive_tracker.py   # Main tracking UI (OpenCV + YOLO)
└── README.md                # Instructions, features, and usage
```

---

## ✅ Features

- ✅ Manual object selection via mouse click
- ✅ Bold & scoped visualization for active tracker
- ✅ Dashed boxes for untracked detections
- ✅ Live terminal feedback (BBox, ID, confidence, center)
- ✅ FPS display toggle
- ✅ Configurable tracker (e.g., bytetrack.yaml)
- ✅ Clean and modular code with inline documentation
- ✅ Companion script to prepare any YOLO model (via download or custom path)

---

## 🎯 Purpose

- To provide a high-quality, educational example showing how to integrate YOLO + tracker + OpenCV with user interaction
- Useful for demos, research, debugging, and UI/UX enhancements

---

## 🛠️ How to Use

1. Run `add_yolo_model.py` to download and export a YOLO model (e.g., yolov5s.pt)
2. Run `interactive_tracker.py` and click on objects to track them in real time

See full instructions in the included `README.md`.

---

## 🔗 Related

No existing issue linked. This is a standalone contribution.

---

## 👤 Author

**Alireza Ghaderi**  
📅 March 2025  
🔗 [[LinkedIn](https://www.linkedin.com/in/alireza787b)](https://www.linkedin.com/in/alireza787b)

---

## 📜 Contributor License Agreement (CLA)

I have read the CLA Document and I sign the CLA ✅

---

Thank you 🙏 for reviewing this contribution. I'm happy to revise or expand based on feedback from the Ultralytics team!


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary  
Introduced a new interactive object tracking UI using YOLO11 and OpenCV, enabling real-time object detection and tracking with user-friendly features. 🚀  

### 📊 Key Changes  
- Added a new example project: **YOLO Interactive Tracking UI**.  
- Features include:  
  - Real-time object detection and tracking.  
  - Click-to-track functionality for selecting objects.  
  - Visual enhancements like bold/dashed bounding boxes and tracking scope lines.  
  - Adjustable tracking algorithms (`bytetrack`, `botsort`).  
  - Compatibility with PyTorch (`.pt`) and NCNN (`.param + .bin`) models for GPU and CPU devices.  
- Provided a detailed README with setup instructions, hardware compatibility, and usage examples.  
- Added a Python script (`interactive_tracker.py`) for running the UI.  

### 🎯 Purpose & Impact  
- **Purpose**:  
  - Simplify object tracking for users with an intuitive interface.  
  - Showcase YOLO11's tracking capabilities in real-time applications.  
- **Impact**:  
  - Enhances user experience with interactive tracking and visualization.  
  - Supports diverse hardware setups, from Raspberry Pi to high-performance GPUs.  
  - Encourages experimentation and integration into custom projects.  

This update is perfect for developers and enthusiasts looking to explore or demonstrate object tracking capabilities interactively. 🖱️✨

alireza787b committed:
added interactive tracker example UI

alireza787b committed:
added demo gif

Signed-off-by: Alireza Ghaderi <p30planets@gmail.com>

alireza787b committed:
Update README.md

Signed-off-by: Alireza Ghaderi <p30planets@gmail.com>

alireza787b committed:
Update README.md

Signed-off-by: Alireza Ghaderi <p30planets@gmail.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

UltralyticsAssistant commented:
👋 Hello @alireza787b, thank you for submitting your pull request to the `ultralytics/ultralytics` repository! 🚀 Your contribution looks fantastic and we appreciate the effort you've put into adding this interactive object tracking example. To ensure a smooth review and integration process, please review the following checklist:

- ✅ **Define a Purpose**: You've already provided a clear purpose and summary in your PR description—great job! If there are any related issues or discussions, consider linking to them for additional context.
- ✅ **Synchronize with Source**: Ensure your branch is up-to-date with the `main` branch of the `ultralytics/ultralytics` repository. If your branch is behind, update it by clicking the 'Update branch' button or by running `git pull` and resolving any conflicts locally.
- ✅ **Ensure CI Checks Pass**: Verify that all [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues and update your PR.
- ✅ **Update Documentation**: You've included a `README.md` file with detailed instructions and examples—excellent work! If additional documentation updates are required, ensure they are included.
- ✅ **Add Tests**: If applicable, consider adding or updating tests to cover your changes. Confirm that all tests pass locally before updating the PR.
- ✅ **Sign the CLA**: We noticed you've acknowledged the [Contributor License Agreement (CLA)](https://docs.ultralytics.com/help/CLA/)—thank you! If this is your first contribution, ensure you've signed the CLA by adding the statement "I have read the CLA Document and I sign the CLA" to your PR comments.
- ✅ **Minimize Changes**: Keep your changes focused and concise. It looks like you've done a great job of this already, ensuring a clear and well-organized contribution. _"Focus is saying no to a hundred good ideas." — Steve Jobs_

For more details on contributing, please refer to the [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If you have any questions or need assistance, feel free to leave a comment. An Ultralytics engineer will review your PR and provide feedback soon. Thank you for your contribution to Ultralytics! 🌟

alireza787b commented:
I have read the CLA Document and I sign the CLA

alireza787b commented:
recheck

alireza787b commented:
I have read the CLA Document and I sign the CLA

alireza787b committed:
currected docs

alireza787b committed:
Merge branch 'main' of https://github.com/alireza787b/ultralytics

alireza787b committed:
fixed final dco

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

alireza787b committed:
Merge branch 'main' into main

RizwanMunawar commented:
Hi @alireza787b, thanks for the PR! I’ll need a bit of time to review it. Overall implementation is fine, we will need to adjust some visualization functions!. Thanks

alireza787b commented:
> Hi @alireza787b, thanks for the PR! I’ll need a bit of time to review it. Overall implementation is fine, we will need to adjust some visualization functions!. Thanks

Hi @RizwanMunawar 
Thanks..
I am going to build a youtube short video for this as well for education .. so I will wait until it merges 
Regards

alireza787b committed:
Merge branch 'main' into main

alireza787b committed:
gpu smart trcker test

alireza787b committed:
Merge branch 'main' of https://github.com/alireza787b/ultralytics

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

alireza787b committed:
update with gpu readme

alireza787b committed:
Merge branch 'main' of https://github.com/alireza787b/ultralytics

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

alireza787b committed:
updted docs

alireza787b committed:
Merge branch 'main' of https://github.com/alireza787b/ultralytics

glenn-jocher committed:
Merge branch 'main' into main

alireza787b committed:
uses yolo11 instead

alireza787b committed:
Merge branch 'main' of https://github.com/alireza787b/ultralytics

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

alireza787b committed:
Merge branch 'main' into main

RizwanMunawar committed:
update readme.md

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

RizwanMunawar committed:
update README.md

glenn-jocher committed:
Merge branch 'main' into main

alireza787b committed:
bypass weight only in pytorch 2.6 

Signed-off-by: Alireza Ghaderi <p30planets@gmail.com>

alireza787b committed:
Merge branch 'ultralytics:main' into main

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

RizwanMunawar commented:
@alireza787b Hi, I have reviewed your PR Please check my above comments. Overall, I don't think we need to `add_yolo_model.py`  file.

Please let me know once the changes are done, and I can review them again. Thanks

alireza787b committed:
removed add_model

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

alireza787b committed:
thicker line

alireza787b committed:
Merge branch 'main' of https://github.com/alireza787b/ultralytics into main

alireza787b committed:
disclaimer

alireza787b commented:
Hi @RizwanMunawar , thanks for the review!
I didn’t see any inline comments, but I understood your point regarding add_yolo_model.py. I absolutely agree.
I've now removed the file entirely, and updated the interactive_tracker.py docstring and README.md accordingly:

No longer referencing or depending on add_yolo_model.py

Clarified model loading (users can now specify paths manually)

Removed the yolo/ directory structure requirement

Improved clarity for model selection (PyTorch vs NCNN)

Let me know if you spot anything else or would like further adjustments. Appreciate your time!

alireza787b committed:
Merge branch 'main' into main

alireza787b committed:
pt file default

RizwanMunawar commented:
@alireza787b Please see here!
![Screenshot 2025-04-01 235141](https://github.com/user-attachments/assets/2b217f68-99f1-41e6-9349-bbf0e8981a12)

RizwanMunawar commented:
Also, I updated the readme in the commit: https://github.com/ultralytics/ultralytics/pull/19854/commits/86e155626ca93a14e28140eb454c9c66f5a3dc7d

Let's please use this, as it directly aligns with Ultralytics workflows. Thanks

alireza787b committed:
used antive color class

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

alireza787b committed:
more readbale label

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

alireza787b committed:
Merge branch 'main' into main

alireza787b committed:
udpaed readme templates and docs

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

alireza787b committed:
Merge branch 'main' into main

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

RizwanMunawar committed:
Merge branch 'main' into main

RizwanMunawar committed:
update README.md

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

RizwanMunawar committed:
update README.md

RizwanMunawar committed:
update README.md

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

RizwanMunawar committed:
update README.md

RizwanMunawar committed:
Merge branch 'main' of https://github.com/alireza787b/ultralytics

RizwanMunawar committed:
update README.md

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

RizwanMunawar committed:
update README.md

RizwanMunawar committed:
Merge branch 'main' of https://github.com/alireza787b/ultralytics

RizwanMunawar committed:
update README.md

RizwanMunawar committed:
update README.md

RizwanMunawar committed:
update README.md

RizwanMunawar committed:
update README.md

RizwanMunawar committed:
update interactive_tracker.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

RizwanMunawar committed:
update interactive_tracker.py

RizwanMunawar committed:
update interactive_tracker.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

RizwanMunawar committed:
update interactive_tracker.py

RizwanMunawar committed:
update interactive_tracker.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

RizwanMunawar commented on examples/YOLO-Interactive-Tracking-UI/interactive_tracker.py:
```diff
@@ -0,0 +1,206 @@
+        color = object_colors.setdefault(track_id, tuple(reversed(colors(track_id, False))))
+        label = f"{model.names[class_id]} ID {track_id} ({conf:.2f})"
+
+        if track_id == selected_object_id:
```
@alireza787b Hi, I have reviewed again! Made some changes. Not sure why we are using the opencv functions, we already have these in [plotting.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/plotting.py), can we please use these? Thanks

alireza787b replied:
Thanks for the follow-up @RizwanMunawar!

You're absolutely right — Ultralytics' `plotting.py` (especially the `Annotator` and `colors` utilities) is the preferred way, and I've now updated the code accordingly to use `Annotator.box_label()` and `colors(...)` for all core drawing (labels and bounding boxes).

That said, I originally used OpenCV for a few UX-focused enhancements like:
- scope lines from bbox edges (to give a lock-on feel),
- a center point marker, and
- a pulsing highlight on the selected object.

These were just meant to improve visual clarity and interactivity, especially for beginners. I’ve retained them *only* as additive UX layers — not as replacements for any built-in Ultralytics utilities.

Let me know if you'd prefer I remove these enhancements too — but currently, the main drawing is fully aligned with Ultralytics `plotting.py` standards. Or we can revert to the last commit and use the custom scope and dashed style..
 Thanks again!

RizwanMunawar commented on examples/YOLO-Interactive-Tracking-UI/interactive_tracker.py:
```diff
@@ -0,0 +1,206 @@
+    cv2.line(im, mid_right, extend_line_from_edge(*mid_right, "right", im.shape), color, 2)
+
+
+def draw_label_with_bg(img, text, position, color, font_scale=0.6, thickness=1):
```
@alireza787b, we are not using this function in more than 1 place, let's please remove this function! Thanks

alireza787b replied:
I've now removed the unused `draw_label_with_bg()` function entirely.

Appreciate the cleanup feedback!

RizwanMunawar reviewed (changes requested):
@alireza787b Hi, I have reviewed the PR, please make the changes I requested. Let me know once changes done and I can review it again. Thanks

alireza787b committed:
removed scope and opencv drawings

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

RizwanMunawar committed:
cleanup

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

RizwanMunawar committed:
update README.md

RizwanMunawar committed:
add video writer functionality

RizwanMunawar committed:
Merge branch 'main' into main

alireza787b committed:
tested locally, added flag for save video control

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com

alireza787b committed:
Update demo

Signed-off-by: Alireza Ghaderi <p30planets@gmail.com>

alireza787b committed:
Merge branch 'main' into main

RizwanMunawar committed:
Merge branch 'main' into main

RizwanMunawar commented:
Hey @alireza787b, thanks for updating the GIF! Do you happen to have the raw video file? If you can share it with me, that would be awesome. I need to compress it, since the current version is 7MB, which is a bit too large for a README. We usually try to keep videos or images under 1 MB. Thanks

RizwanMunawar committed:
cleanup

RizwanMunawar committed:
cleanup

alireza787b commented:
@RizwanMunawar 
Sure. below is the raw video (compressed to 10 MB) 

https://github.com/user-attachments/assets/fe5e743d-9685-4547-b3b6-0bc7935b8f43

RizwanMunawar commented:
@alireza787b ok got it. Thanks!

How about this video? I think it can highlight the idea properly. What do you think?

https://github.com/user-attachments/assets/723e919e-555b-4cca-8e60-18e711d4f3b2

RizwanMunawar committed:
update readme.md

RizwanMunawar committed:
update readme.md

RizwanMunawar committed:
update readme.md

RizwanMunawar reviewed (approved):
@alireza787b! Nice work on this. looks good.

alireza787b commented:
Yeah .this demo is also nice 
 That knew just tried to emphasize on dynamic moving camera and show the cursor ...
 But this one is also nice 
No problem that's good 
 I will also build a YouTube video demo of this late maybe we add that YouTube as well 

Thanks for the help

glenn-jocher committed:
Merge branch 'main' into main

glenn-jocher commented:
@alireza787b PR merged, thank you for your contributions!

UltralyticsAssistant commented:
🎉 **Merged and Magnificent!**  

"Alone we can do so little; together we can do so much." – Helen Keller  

A huge thank you to @alireza787b for leading the charge on this innovative interactive tracking UI, and to @glenn-jocher and @RizwanMunawar for your invaluable contributions. This collaboration has unlocked new possibilities for YOLO11, blending cutting-edge tech with user-friendly design to inspire developers worldwide. 🚀  

Your work not only simplifies real-time object tracking but also exemplifies the power of teamwork and creativity. Here's to pushing boundaries and making AI more accessible! 👏✨
